packages/core/js-client/src/compilerSupport/registerService.ts

```diff
@@ -13,43 +13,76 @@
-import { logger } from '../util/logger.js';
-
-const log = logger('aqua');
-
-export const registerService: RegisterServiceType = ({ peer, def, serviceId, service }) => {
-    log.trace('registering aqua service %o', { def, serviceId, service });
-
-    // Checking for missing keys
-    const requiredKeys = def.functions.tag === 'nil' ? [] : Object.keys(def.functions.fields);
-    const incorrectServiceDefinitions = requiredKeys.filter((f) => !(f in service));
-    if (!!incorrectServiceDefinitions.length) {
-        throw new Error(
-            `Error registering service ${serviceId}: missing functions: ` +
-                incorrectServiceDefinitions.map((d) => "'" + d + "'").join(', '),
-        );
-    }
-
-    if (!serviceId) {
-        serviceId = def.defaultServiceId;
-    }
-
-    if (!serviceId) {
-        throw new Error('Service ID must be specified');
-    }
-
-    const singleFunctions = def.functions.tag === 'nil' ? [] : Object.entries(def.functions.fields);
-    for (let singleFunction of singleFunctions) {
-        let [name, type] = singleFunction;
-        // The function has type of (arg1, arg2, arg3, ... , callParams) => CallServiceResultType | void
-        // Account for the fact that user service might be defined as a class - .bind(...)
-        const userDefinedHandler = service[name].bind(service);
-
-        const serviceDescription = userHandlerService(serviceId, singleFunction, userDefinedHandler);
-        registerGlobalService(peer, serviceDescription);
-    }
-    log.trace('aqua service registered %s', serviceId);
+
+import type { ServiceDef, ServiceImpl } from "@fluencelabs/interfaces";
+
+import { FluencePeer } from "../jsPeer/FluencePeer.js";
+import { logger } from "../util/logger.js";
+
+import { registerGlobalService, userHandlerService } from "./services.js";
+
+const log = logger("aqua");
+
+interface RegisterServiceArgs {
+  peer: FluencePeer;
+  def: ServiceDef;
+  serviceId: string | undefined;
+  service: ServiceImpl;
+}
+
+export const registerService = ({
+  peer,
+  def,
+  serviceId,
+  service,
+}: RegisterServiceArgs) => {
+  // TODO: Need to refactor this. We can compute function types from service implementation, making func more type safe
+  log.trace("registering aqua service %o", { def, serviceId, service });
+
+  // Checking for missing keys
+  const requiredKeys =
+    def.functions.tag === "nil" ? [] : Object.keys(def.functions.fields);
+
+  const incorrectServiceDefinitions = requiredKeys.filter((f) => {
+    return !(f in service);
+  });
+
+  if (incorrectServiceDefinitions.length > 0) {
+    throw new Error(
+      `Error registering service ${serviceId}: missing functions: ` +
+        incorrectServiceDefinitions
+          .map((d) => {
+            return "'" + d + "'";
+          })
+          .join(", "),
+    );
+  }
+
+  if (serviceId == null) {
+    serviceId = def.defaultServiceId;
+  }
+
+  if (serviceId == null) {
+    throw new Error("Service ID must be specified");
+  }
```

Comment:
Will this work? Also wanna move the check above so erro messages never include `undefined`
```suggestion
  serviceId = def.defaultServiceId,
  service,
}: RegisterServiceArgs) => {
  // TODO: Need to refactor this. We can compute function types from service implementation, making func more type safe
  log.trace("registering aqua service %o", { def, serviceId, service });

  // Checking for missing keys
  const requiredKeys =
    def.functions.tag === "nil" ? [] : Object.keys(def.functions.fields);

  const incorrectServiceDefinitions = requiredKeys.filter((f) => {
    return !(f in service);
  });

  if (serviceId == null) {
    throw new Error("Service ID must be specified");
  }

  if (incorrectServiceDefinitions.length > 0) {
    throw new Error(
      `Error registering service ${serviceId}: missing functions: ` +
        incorrectServiceDefinitions
          .map((d) => {
            return "'" + d + "'";
          })
          .join(", "),
    );
  }
```